enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/BenchmarkProcedure.java

```diff
@@ -0,0 +1,157 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.replication;
+
+import java.io.IOException;
+import java.util.concurrent.Future;
+import java.util.stream.Stream;
+
+import org.neo4j.causalclustering.core.state.machines.dummy.DummyRequest;
+import org.neo4j.kernel.api.exceptions.InvalidArgumentsException;
+import org.neo4j.logging.Log;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.Procedure;
+
+import static java.lang.Math.toIntExact;
+import static org.neo4j.procedure.Mode.DBMS;
+
+@SuppressWarnings( "unused" )
+public class BenchmarkProcedure
+{
+    @Context
+    public Replicator replicator;
+
+    @Context
+    public Log log;
+
+    private static long startTime;
+    private static Worker[] workerArr;
+
+    @Description( "Start the benchmark." )
+    @Procedure( name = "dbms.cluster.benchmark.start", mode = DBMS )
+    public synchronized void start( @Name( "nThreads" ) Long nThreads, @Name( "blockSize" ) Long blockSize ) throws InvalidArgumentsException, IOException
+    {
+        if ( workerArr != null )
+        {
+            throw new IllegalStateException( "Already running." );
+        }
+
+        log.info( "Starting replication benchmark procedure" );
+
+        startTime = System.currentTimeMillis();
+        workerArr = new Worker[toIntExact( nThreads )];
+
+        for ( int i = 0; i < nThreads; i++ )
+        {
+            workerArr[i] = new Worker( toIntExact( blockSize ) );
+            workerArr[i].start();
+        }
+    }
+
+    @Description( "Stop a running benchmark." )
+    @Procedure( name = "dbms.cluster.benchmark.stop", mode = DBMS )
+    public synchronized Stream<BenchmarkResult> stop() throws InvalidArgumentsException, IOException, InterruptedException
+    {
+        if ( workerArr == null )
+        {
+            throw new IllegalStateException( "Not running." );
+        }
+
+        log.info( "Stopping replication benchmark procedure" );
+
+        for ( Worker worker : workerArr )
+        {
+            worker.stop();
+        }
+
+        for ( Worker worker : workerArr )
+        {
+            worker.join();
+        }
+
+        long runTime = System.currentTimeMillis() - startTime;
+
+        long totalRequests = 0;
+        long totalBytes = 0;
+
+        for ( Worker worker : workerArr )
+        {
+            totalRequests += worker.totalRequests;
```

Comment:
Or `long totalRequests = Arrays.stream ( workerArr ).mapToLong( worker -> worker.totalRequests ).sum()` which wouldn't require changes to `Worker`.

Stream all the things! :smile: